tests/conftest.py

```diff
@@ -665,7 +665,7 @@ def ed25519_node_keys(node_keys):
     return node_keys_dict
 
 
-@pytest.fixture(scope='session')
+@pytest.fixture
```

Comment:
Do we need it loaded for the whole test session? I think its better to keep a few fixtures in place as possible, to minimize the possibility for cross contamination between tests.

Consider the case where we write a test in future that loads the node keys, changes a private key in the file, and tests to make sure we detect the mismatched public/private pair. Scoped for the session, we'd then have that dirty pair loaded for all tests that run afterwards.

Reply:
> Consider the case where we write a test in future that loads the node keys, changes a private key in the file, and tests to make sure we detect the mismatched public/private pair. Scoped for the session, we'd then have that dirty pair loaded for all tests that run afterwards.

This is a completely hypothetical case which I am not really sure when will it happen. Could you please elaborate when will such a situation occur with respect to our current codebase or near term feature set?

Reply:
It isn't a problem in the current code base. But as a general practice, we shouldn't scope things to `session` unless we have an explicit use case. It's not good testing practice, which is why the `scope` defaults to single tests, and why you need to specify `session` as a value.

Since nothing breaks when we remove the `session` level scoping, we don't actually need it, and it makes the tests dirtier.

Reply:
Can you please give an example of how the `node_keys` fixture can become dirty? Its response is static so I want to understand how will it become dirty.

Reply:
> It's not good testing practice, which is why the scope defaults to single tests, and why you need to specify session as a value.

Could please point me to where did you come across this explanation in the pytest docs?

Reply:
I've already come up with a hypothetical scenario where this could be a problem.

Let me turn the question around. What is the argument for keeping this scoped as `session`? I make the following argument:

1) Smaller code base is better than larger. Since the tests run fine without this parameter, we should remove it
2) We already have a problem with tests bleeding into one another. `session` level scoping only increases this risk. If we don't need data to be shared between tests, we shouldn't share data between tests.

Is there a good argument for scoping this as session that I'm missing?

Reply:
> I've already come up with a hypothetical scenario where this could be a problem.

> Can you please give an example of how the node_keys fixture can become dirty? Its response is static so I want to understand how will it become dirty.

I was asking for the code of this hypothetical scenario. I would really help if you can find such an example.

> 1. Smaller code base is better than larger. Since the tests run fine without this parameter, we should remove it

The line is about 32 characters, I don't really see how is that any reduction or increase in the size of the code base. Our code is more than 10K+ Loc adding or subtracting 1 to that doesn't really have any impact there.

> 2. We already have a problem with tests bleeding into one another. session level scoping only increases this risk. If we don't need data to be shared between tests, we shouldn't share data between tests.

The tests you are talking about is `abci` tests and that is the point of these tests. They are supposed to bleed into one another. They have to share state.

> Is there a good argument for scoping this as session that I'm missing?

My argument is pretty simple i.e. the fixture returns a static value so there is no point re-evaluating the fixture every time its called, so we can scope it to the session. But since you [mentioned](https://github.com/bigchaindb/bigchaindb/pull/2458#discussion_r212280422) that this a bad practice and the reason why we don't have `session` as the default scope, I just wanted to know where the `pytest` docs state that and understand it better.

Reply:
I'm not finding anything online to state one way or another how we should scope this.

However, speaking with a few years of experience as a test engineer, my experience tells me that cross contamination on the unit and integration test level is bad news.

We want tests that, in the event of failure, give us clear indicators of what is broken. This means tests should cover a specific interaction in the code, as narrowly as possible, with as few side effects as possible. This way, if something turns red, we know exactly where to look.

If we want to check for side effects that happen between code paths, those deserve their own tests. We shouldn't be checking for them as a 'by product' of normal test execution. The reason for this is two fold. First, with dedicated tests for cross contamination, we get better coverage of corner cases. As it stands, we have a relatively static order of execution on a deterministic system, so while it may seem that we're covering 'weird db states', we really only cover one random walk through our product, which we do over and over. Much better for exploratory testing to have something like ChaosMonkey, which does some truly exploratory testing.

Second, as it stands now, even if we do uncover a weird interaction due to broad scoping, our test reporting isn't designed in a way that would allow for a meaningful investigation. If I (or someone else) did run into a case where a session level fixture caused a test to break, the most likely result would just be to down-scope the fixture and move on. With overly broad, poorly defined tests, there is no goo way to tell when a failure is due to a broken test, or due to broken code. Again, if we want to test for weird cross interactions in our code (a generally good idea!) we should have some dedicated tests that do this.

Finally, as a further argument against test overlap, I've already run into the situation once this week where a problem in Test A caused a failure in Test Q (somewhere far away in the test execution order.) This was a debugging nightmare, and cost me about a day of time to dig out the problem. In the end, it was exactly the problem where fixtures didn't clean up after themselves that caused the issue (in this case, changing the validator set and not restoring it after the test ended.)

To reiterate, I agree that we should be testing for weird interactions, but I think this is a better candidate for scenario style acceptance tests. What we're writing here is somewhere between a unit and an integration test. In that context, my experience has been that over complicated tests lead to nightmare debugging when something breaks, and cost more time than the fraction of a second it saves to share fixtures between tests.

Reply:
You're right, I am generalizing the scope of the discussion, because I'm trying to have a conversation about best practices in testing, and I'm making the argument that as a general rule, the scope of testing assets should be as small as possible while still allowing the test to cover it's intended target.

Given that I think that's a good rule of thumb, and given that no change in coverage occurs when I remove the `session` level scoping from this fixture, I consider this to be an improvement to our code, in the same way that removing unused imports is an improvement. Even if an unused import doesn't break anything, per se, it's still a good idea to clean them up when you see them.

If you have a good counter argument to my rule of thumb, I'm open to hearing it, but if not, I think it justifies this change to the code.

Reply:
> If you have a good counter argument to my rule of thumb, I'm open to hearing it, but if not, I think it justifies this change to the code.

Well, my argument is pretty straightforward, I don't see any way a static fixture can violate your rule of thumb. Your rule of thumb is derived from the experience that larger scopes can end up contaminating state, what I am saying is that `node_keys` is static and will never contaminate state so it satisfies the underlying reasoning for your rule of thumb.